BackEnd/chat_notifications/consumer.py

```
@@ -0,0 +1,72 @@
+from channels.generic.websocket import AsyncWebsocketConsumer
+import json
+import logging
+from channels.exceptions import ChannelFull
+
+logger = logging.getLogger(__name__)
+
+
+class NotificationConsumer(AsyncWebsocketConsumer):
+    async def connect(self):
+        """Handle WebSocket connection."""
+        if not self.scope["user"].is_authenticated:
+            logger.warning("Unauthenticated user attempted to connect.")
+            await self.close()
+            return
+
+        self.room_group_name = f'notifications_{self.scope["user"].id}'
+
+        if self.channel_layer is None:
+            logger.error("channel_layer is not configured properly.")
+            await self.close()
+            return
+
+        try:
+            await self.channel_layer.group_add(
+                self.room_group_name, self.channel_name
+            )
+            await self.accept()
+            logger.info(
+                f"User {self.scope['user'].id} connected to notifications."
+            )
+        except ValueError as e:
+            logger.error(f"Invalid value when adding user to group: {str(e)}")
+            await self.close()
+        except RuntimeError as e:
+            logger.error(f"Runtime error when adding user to group: {str(e)}")
+            await self.close()
+
+    async def disconnect(self, close_code):
+        """Handle WebSocket disconnection."""
+        if self.channel_layer is not None:
+            try:
+                await self.channel_layer.group_discard(
+                    self.room_group_name, self.channel_name
+                )
+                logger.info(
+                    f"User {self.scope['user'].id} disconnected from notifications."
+                )
+            except KeyError as e:
+                logger.error(f"KeyError in group_discard: {str(e)}")
+            except RuntimeError as e:
+                logger.error(f"Runtime error in group_discard: {str(e)}")
```

If `connect()` fails early (e.g., due to `channel_layer` issues), `disconnect()` may still try to call `group_discard` on an undefined `room_group_name`. This could cause unexpected errors on disconnect.
Ensure `room_group_name` exists before `group_discard`
```
if hasattr(self, "room_group_name") and self.channel_layer is not None:
    try:
        await self.channel_layer.group_discard(self.room_group_name, self.channel_name)
        logger.info(f"User {self.scope['user'].id} disconnected from notifications.")
    except Exception as e:
        logger.error(f"Error during group_discard: {str(e)}")
```

---

fixed